osu.Game/Screens/Edit/HitObjectChangeHandler.cs

```diff
@@ -0,0 +1,293 @@
+    /// </summary>
+    /// <inheritdoc/>
+    public partial class HitObjectChangeHandler : TransactionalCommitComponent
+    {
+        private readonly EditorBeatmap editorBeatmap;
+
+        /// <summary>
+        /// This change handler will be suppressed while a transaction with this command handler is in progress.
+        /// Any save states of this change handler will be added to the undo stack.
+        /// </summary>
+        private readonly EditorChangeHandler? changeHandler;
+
+        public readonly Bindable<bool> CanUndo = new BindableBool();
+
+        public readonly Bindable<bool> CanRedo = new BindableBool();
+
+        public event Action? OnStateChange;
+
+        public bool HasUncommittedChanges => currentTransaction.UndoChanges.Count != 0;
+
+        public Guid CurrentState => undoStack.TryPeek(out var transaction) ? transaction.Id : Guid.Empty;
+
+        private Transaction currentTransaction;
+
+        private readonly Stack<Transaction> undoStack = new Stack<Transaction>();
+
+        private readonly Stack<Transaction> redoStack = new Stack<Transaction>();
+
+        private bool isRestoring;
+
+        public HitObjectChangeHandler(EditorBeatmap editorBeatmap, EditorChangeHandler? changeHandler = null)
+        {
+            currentTransaction = new Transaction();
+            this.editorBeatmap = editorBeatmap;
+            this.changeHandler = changeHandler;
+
+            editorBeatmap.TransactionBegan += BeginChange;
+            editorBeatmap.TransactionEnded += EndChange;
+            editorBeatmap.SaveStateTriggered += SaveState;
+
+            if (this.changeHandler != null)
+                this.changeHandler.OnStateChange += commitChangeHandlerStateChange;
+        }
+
+        /// <summary>
+        /// Submits a change to be applied and added to the history.
+        /// </summary>
+        /// <param name="change">Change to be applied.</param>
+        /// <param name="commitImmediately">Whether to commit the current transaction and push it onto the undo stack immediately.</param>
+        public void Submit(IRevertibleChange change, bool commitImmediately = false)
+        {
+            if (commitImmediately)
+                BeginChange();
+
+            change.Apply();
+            Record(change);
+
+            if (commitImmediately)
+                EndChange();
+        }
+
+        /// <summary>
+        /// Submits a collection of changes to be applied and added to the history.
+        /// </summary>
+        /// <param name="changes">Changes to be applied.</param>
+        /// <param name="commitImmediately">Whether to commit the current transaction and push it onto the undo stack immediately.</param>
+        public void Submit(IEnumerable<IRevertibleChange> changes, bool commitImmediately = false)
+        {
+            foreach (var change in changes)
+                Submit(change);
+
+            if (commitImmediately)
+                UpdateState();
+        }
```

Comment:
- Why does this method exist? Nothing calls it.
- Why are the `changes` not one atomic change but a series of changes? Seems confusing to me, and should be xmldoc'd.
- Why is this calling `UpdateState()` at the end? Should I be worried that the `Submit()` overload that takes a single change doesn't?

Reply:
Might've had some use in the past and is now obsolete. I'll remove it.